data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataDecoder.java

```diff
@@ -0,0 +1,401 @@
+  {
+    try
+    {
+      NonBlockingDataParser.Token token;
+      while ((token = _parser.nextToken()) != EOF_INPUT)
+      {
+        validate(token);
+        switch (token)
+        {
+          case START_OBJECT:
+            _currObjSize = _parser.currentComplexObjSize();
+            if (_currObjSize >= 0)
+            {
+              push(new DataMap(DataMapBuilder.getOptimumHashMapCapacityFromSize(_currObjSize)), false);
+            }
+            else
+            {
+              // If we are already filling out a DataMap, we cannot reuse _currDataMapBuilder and thus
+              // need to create a new one.
+              if (_currDataMapBuilder == null || _currDataMapBuilder.inUse()) {
+                _currDataMapBuilder = new DataMapBuilder();
+              }
+              _currDataMapBuilder.setInUse(true);
+              push(_currDataMapBuilder, false);
+            }
+            break;
+          case END_OBJECT:
+            pop();
+            break;
+          case START_ARRAY:
+            _currObjSize = _parser.currentComplexObjSize();
+            if (_currObjSize >= 0)
+            {
+              push(new DataList(_currObjSize), true);
+            }
+            else
+            {
+              push(new DataList(), true);
+            }
+            break;
+          case END_ARRAY:
+            pop();
+            break;
+          case STRING:
+            if (!_isCurrList && _currField == null) {
+              _currField = _parser.getString();
+              _expectedTokens = VALUE;
+            } else {
+              addValue(_parser.getString());
+            }
+            break;
+          case INTEGER:
+            addValue(_parser.getIntValue());
+            break;
+          case LONG:
+            addValue(_parser.getLongValue());
+            break;
+          case FLOAT:
+            addValue(_parser.getFloatValue());
+            break;
+          case DOUBLE:
+            addValue(_parser.getDoubleValue());
+            break;
+          case BOOL_TRUE:
+            addValue(Boolean.TRUE);
+            break;
+          case BOOL_FALSE:
+            addValue(Boolean.FALSE);
+            break;
+          case NULL:
+            addValue(Data.NULL);
+            break;
+          case NOT_AVAILABLE:
+            readNextChunk();
+            return;
+          default:
+            handleException(new Exception("Unexpected token " + token + " from data parser"));
+        }
+      }
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  protected final void validate(NonBlockingDataParser.Token token)
+  {
+    if (!(token == NOT_AVAILABLE || _expectedTokens.contains(token)))
+    {
+      handleException(new Exception("Expecting " + _expectedTokens + " but got " + token));
+    }
+  }
+
+  private void push(DataComplex dataComplex, boolean isList)
+  {
+    if (_currObjSize >= 0)
+    {
+      _currObjSizeStack.push(_currObjSize);
```

Comment:
I don't think you'd have to keep the stack of current object sizes. You can just use it as a hint provided by the parser for initializing the datacomplex objects. After that adding/ending the datacomplex should be done only by the parser tokens.